File: packages/engine/src/cards.ts
/** Rank: 2=deuce … 14=Ace. Suit: c d h s */
export type Suit = 'c' | 'd' | 'h' | 's';
export type Rank = 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12 | 13 | 14;

export interface Card {
  rank: Rank;
  suit: Suit;
}

export const SUITS: Suit[] = ['c', 'd', 'h', 's'];
export const RANKS: Rank[] = [2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14];

const RANK_CHARS: Record<string, Rank> = {
  '2': 2,
  '3': 3,
  '4': 4,
  '5': 5,
  '6': 6,
  '7': 7,
  '8': 8,
  '9': 9,
  T: 10,
  J: 11,
  Q: 12,
  K: 13,
  A: 14,
};

const RANK_TO_CHAR: Record<number, string> = {
  2: '2',
  3: '3',
  4: '4',
  5: '5',
  6: '6',
  7: '7',
  8: '8',
  9: '9',
  10: 'T',
  11: 'J',
  12: 'Q',
  13: 'K',
  14: 'A',
};

export function cardToString(card: Card): string {
  return `${RANK_TO_CHAR[card.rank]}${card.suit}`;
}

export function parseCard(s: string): Card {
  if (s.length !== 2) throw new Error(`Invalid card: ${s}`);
  const rank = RANK_CHARS[s[0]!.toUpperCase()];
  const suit = s[1]!.toLowerCase() as Suit;
  if (!rank || !SUITS.includes(suit)) throw new Error(`Invalid card: ${s}`);
  return { rank, suit };
}

export function parseCards(s: string): Card[] {
  return s.trim().split(/\s+/).filter(Boolean).map(parseCard);
}

export function createDeck(): Card[] {
  const deck: Card[] = [];
  for (const suit of SUITS) {
    for (const rank of RANKS) {
      deck.push({ rank, suit });
    }
  }
  return deck;
}

/** Fisher–Yates with CSPRNG (Node crypto or Web Crypto). */
export function shuffle<T>(items: T[], randomBytes: (n: number) => Uint8Array): T[] {
  const arr = items.slice();
  for (let i = arr.length - 1; i > 0; i--) {
    const buf = randomBytes(4);
    const rand = (buf[0]! << 24) | (buf[1]! << 16) | (buf[2]! << 8) | buf[3]!;
    const j = (rand >>> 0) % (i + 1);
    const tmp = arr[i]!;
    arr[i] = arr[j]!;
    arr[j] = tmp;
  }
  return arr;
}

File: apps/android/core/engine/src/main/java/com/felt/android/engine/Cards.kt
package com.felt.android.engine

/** Rank: 2 = deuce … 14 = Ace. Suit: c d h s */
typealias Rank = Int

enum class Suit(val char: Char) {
    CLUBS('c'),
    DIAMONDS('d'),
    HEARTS('h'),
    SPADES('s');

    companion object {
        fun fromChar(c: Char): Suit? = entries.find { it.char == c.lowercaseChar() }
    }
}

data class Card(
    val rank: Rank,
    val suit: Suit,
)

val SUITS: List<Suit> = Suit.entries
val RANKS: List<Rank> = (2..14).toList()

private val RANK_CHARS: Map<Char, Rank> = mapOf(
    '2' to 2, '3' to 3, '4' to 4, '5' to 5, '6' to 6, '7' to 7, '8' to 8, '9' to 9,
    'T' to 10, 'J' to 11, 'Q' to 12, 'K' to 13, 'A' to 14,
)

private val RANK_TO_CHAR: Map<Rank, Char> = RANK_CHARS.entries.associate { (k, v) -> v to k }

fun cardToString(card: Card): String = "${RANK_TO_CHAR[card.rank]}${card.suit.char}"

fun parseCard(s: String): Card {
    require(s.length == 2) { "Invalid card: $s" }
    val rank = RANK_CHARS[s[0].uppercaseChar()]
    val suit = Suit.fromChar(s[1])
    require(rank != null && suit != null) { "Invalid card: $s" }
    return Card(rank, suit)
}

fun parseCards(s: String): List<Card> =
    s.trim().split(Regex("\\s+")).filter { it.isNotEmpty() }.map(::parseCard)

fun createDeck(): List<Card> = buildList {
    for (suit in SUITS) {
        for (rank in RANKS) {
            add(Card(rank, suit))
        }
    }
}

/** Fisher–Yates shuffle using the supplied CSPRNG byte provider. */
fun <T> shuffle(items: List<T>, randomBytes: (Int) -> ByteArray): List<T> {
    val arr = items.toMutableList()
    for (i in arr.lastIndex downTo 1) {
        val buf = randomBytes(4)
        require(buf.size >= 4) { "randomBytes must return at least 4 bytes" }
        val rand = ((buf[0].toInt() and 0xFF) shl 24) or
            ((buf[1].toInt() and 0xFF) shl 16) or
            ((buf[2].toInt() and 0xFF) shl 8) or
            (buf[3].toInt() and 0xFF)
        val j = (rand.toUInt() % (i + 1).toUInt()).toInt()
        val tmp = arr[i]
        arr[i] = arr[j]
        arr[j] = tmp
    }
    return arr
}
